doc/howtos/rdtraining/constraints.rst

```diff
@@ -0,0 +1,128 @@
+users from encoding incorrect data? For example, in our real estate module, nothing prevents the
+user to set a negative expected price.
+
+Odoo provides two ways to set up automatically verified invariants:
+:func:`Python constraints <odoo.api.constrains>` and
+:attr:`SQL constraints <odoo.models.Model._sql_constraints>`.
+
+SQL
+===
+
+**Reference**: the documentation related to this topic can be found in
+:ref:`reference/orm/models` and in the `PostgreSQL's documentation`_.
+
+.. note::
+
+    **Goal**: at the end of this section:
+
+    - Amounts should be (strictly) positive
+
+    .. image:: constraints/media/sql_01.gif
+        :align: center
+        :alt: Constraints on amounts
+
+    - Property types and tags have a unique name
+
+    .. image:: constraints/media/sql_02.gif
+        :align: center
+        :alt: Constraints on names
+
+SQL constraints are defined through the model attribute
+:attr:`~odoo.models.Model._sql_constraints`. The latter is assigned to a list
+of triples of strings ``(name, sql_definition, message)``, where ``name`` is a
+valid SQL constraint name, ``sql_definition`` is a table_constraint_ expression,
+and ``message`` is the error message.
+
+You can find a simple example
+`here <https://github.com/odoo/odoo/blob/24b0b6f07f65b6151d1d06150e376320a44fd20a/addons/analytic/models/analytic_account.py#L20-L23>`__.
+
+.. exercise:: Add SQL constraints
+
+    Add the following constraints in the corresponding models:
+
+    - A property expected price must be strictly positive
+    - A property selling price must be positive
+    - An offer price must be strictly positive
+    - A property tag name and property type name must be unique
+
+    Tip: search for the ``unique`` keyword in the Odoo codebase for the name uniqueness.
+
+Restart the server with the ``-u estate`` option to see the result. Note that you might have data
+preventing an SQL constraint to be set. such error message might pop up.
+
+.. code-block:: text
+
+    ERROR rd-demo odoo.schema: Table 'estate_property_offer': unable to add constraint 'estate_property_offer_check_price' as CHECK(price > 0)
+
+For example, if some offers have a price of zero, the constraint can't be applied. You can delete
+the problematic data in order to apply the new constraints.
+
+Python
+======
+
+**Reference**: the documentation related to this topic can be found in
+:func:`~odoo.api.constrains`.
+
+.. note::
+
+    **Goal**: at the end of this section, it is not possible to accept an offer with a price
+    lower than 90% of the expected price.
+
+    .. image:: constraints/media/python.gif
+        :align: center
+        :alt: Python constraint
+
+SQL constraints are an efficient way of ensuring data consistency. However, such constrains cannot
+be 'dynamic' in the sense that cross-checking data might not be possible. In our real estate module
+we would like to prevent a selling price lower than 90% of the expected price. To do this, we
+need a Python constraint.
```

Comment:
Maybe change this paragraph to be more generic. "Some times sql constraints are not enough, and we need to use python to prevent users from encoding incorrect data"